### ISSUE TITLE
Vector type is lost after serialising and deserialising a vector (just a table)

### ISSUE THREAD

**Issue by Zughy** [minetest/minetest#14357]
### Minetest version

```true
5.8.0 (but I guess it's always been like this)
```

### Operating system and version

Pop!_OS

### Summary

Would it be possible to keep the vector nature when de/serialising vectors?

### Steps to reproduce

```lua
local vct = vector.new(0, 30, 1)

minetest.register_on_joinplayer(function(player)
	local vct_str = minetest.serialize(vct)
	
	minetest.chat_send_all("vct_str = " .. vct_str )
	
	vct = minetest.deserialize(vct_str)
	
	minetest.chat_send_all("is vct a vector = " .. tostring(vector.check(vct)))
end)

--[[ output:
vct_str = return {y=30,z=1,x=0}
is vct a vector = false
```

**Comment by appgurueu** [comment]
> 5.8.0 (but I guess it's always been like this)

Yes, this is a known issue / limitation. It applies to all metatables. "Metatables" are usually somewhere between data and code, so dealing with them properly is tricky; you usually can't (and won't want to) serialize the code properly, but you can try to remember the "type" of an object to let you restore the metatable.

I'm not sure whether I'd consider it a bug; it's more of a missing feature / convenience. The limitation should probably at least be documented.

> Would it be possible to keep the vector nature when de/serialising vectors?

Yes. I have an implementation of this [in modlib](https://github.com/appgurueu/modlib/blob/master/minetest/luon.lua). It "knows" a few metatables "by name", and emits code like `return ItemStack"<itemstring>"` or `return blah{...}`, providing the necessary constructors in the environment when loading the code.

For vectors, we'll probably want a special case of `return vec(x, y, z)`, where `vec` in the environment is `vector.new`.

I could extend `minetest.(de)serialize` to support this, at least for some (de)serializable built-in types, and maybe providing an interface to let modders register their own metatables.

**Comment by Zughy** [comment]
> I'm not sure whether I'd consider it a bug; it's more of a missing feature / convenience 

Considering that de/serialising functions are MT functions (`minetest.serialize`) and that we implement vectors, I expect that Minetest supports its own implementation. Hence the bug

**Comment by sfan5** [comment]
This is indeed tricky and I'm not sure it's a "bug". When the vector metatable was added people probably knew that serialization wouldn't deal with them.

worth considering: A property you easily lose when fixing this that your serialization generates a result readable by plain Lua.
(also relevant if we expect older MT to read it).

**Comment by y5nw** [comment]
Maybe a simple solution would be to allow mods (and builtin) to register metatables.

Tables with a "recognized" (-> registered) metatable would be serialized as `D({...}, name)`, where `name` is (ideally) a string, and deserialization would take place with `D` being a function that sets the metatable based on the name. A wrapper could additionally be written to the serialized data to allow the data to remain readable in older versions and in plain Lua, such as
```lua
if not D then
  D = function(tbl)
    return tbl
  end
end
```

**Comment by The4codeblocks** [comment]
> For vectors, we'll probably want a special case of `return vec(x, y, z)`, where `vec` in the environment is `vector.new`.

ideally not, otherwise added fields get removed
